google/cloud/opentelemetry/internal/monitored_resource.cc

```diff
@@ -0,0 +1,222 @@
+};
+
+class MonitoredResourceProvider {
+ public:
+  MonitoredResourceProvider(
+      std::string type, std::unordered_map<std::string, OTelKeyMatch> lookup)
+      : type_(std::move(type)), lookup_(std::move(lookup)) {}
+
+  MonitoredResource Process(
+      opentelemetry::sdk::resource::ResourceAttributes const& attributes) {
+    MonitoredResource mr;
+    mr.type = type_;
+    for (auto const& kv : lookup_) {
+      auto found = false;
+      for (auto const& otel_key : kv.second.otel_keys) {
+        auto p = attributes.find(otel_key);
+        if (p != attributes.end()) {
+          found = true;
+          mr.labels[kv.first] = AsString(p->second);
+          break;
+        }
+      }
+      if (!found && kv.second.fallback)
+        mr.labels[kv.first] = *kv.second.fallback;
+    }
+    return mr;
+  }
+
+ private:
+  std::string type_;
+  std::unordered_map<std::string, OTelKeyMatch> lookup_;
+};
+
+MonitoredResourceProvider GceInstance() {
+  return MonitoredResourceProvider("gce_instance",
+                                   {
+                                       {"zone", {{sc::kCloudAvailabilityZone}}},
+                                       {"instance_id", {{sc::kHostId}}},
+                                   });
+}
+
+MonitoredResourceProvider K8sContainer() {
+  return MonitoredResourceProvider(
+      "k8s_container",
+      {
+          {"location", {{sc::kCloudAvailabilityZone, sc::kCloudRegion}}},
+          {"cluster_name", {{sc::kK8sClusterName}}},
+          {"namespace_name", {{sc::kK8sNamespaceName}}},
+          {"pod_name", {{sc::kK8sPodName}}},
+          {"container_name", {{sc::kK8sContainerName}}},
+      });
+}
+
+MonitoredResourceProvider K8sPod() {
+  return MonitoredResourceProvider(
+      "k8s_pod",
+      {
+          {"location", {{sc::kCloudAvailabilityZone, sc::kCloudRegion}}},
+          {"cluster_name", {{sc::kK8sClusterName}}},
+          {"namespace_name", {{sc::kK8sNamespaceName}}},
+          {"pod_name", {{sc::kK8sPodName}}},
+      });
+}
+
+MonitoredResourceProvider K8sNode() {
+  return MonitoredResourceProvider(
+      "k8s_node",
+      {
+          {"location", {{sc::kCloudAvailabilityZone, sc::kCloudRegion}}},
+          {"cluster_name", {{sc::kK8sClusterName}}},
+          {"node_name", {{sc::kK8sNodeName}}},
+      });
+}
+
+MonitoredResourceProvider K8sCluster() {
+  return MonitoredResourceProvider(
+      "k8s_cluster",
+      {
+          {"location", {{sc::kCloudAvailabilityZone, sc::kCloudRegion}}},
+          {"cluster_name", {{sc::kK8sClusterName}}},
+      });
+}
+
+MonitoredResourceProvider GaeInstance() {
+  return MonitoredResourceProvider(
+      "gae_instance",
+      {
+          {"location", {{sc::kCloudAvailabilityZone, sc::kCloudRegion}}},
+          {"module_id", {{sc::kFaasName}}},
+          {"version_id", {{sc::kFaasVersion}}},
+          {"instance_id", {{sc::kFaasInstance}}},
+      });
+}
+
+MonitoredResourceProvider AwsEc2Instance() {
```

Comment:
I may have friends at Azure... Why no support for them?

Reply:
Because I am only capable of copying the work that was done in golang, which does not include Azure.

Reply:
Please file a feature request (in google-cloud-cpp).  It is possible we won't get to it, but we should have a bug to point to.

Reply:
Okay.  One of the reasons to use OTel is that our customers may be sending the tracing / monitoring data to something other than Cloud Monitoring.

Reply:
Yeah, this PR is to support a special mapping from OTel conventions to Cloud Monitoring conventions.

The spec calls for doing this translation, and also dumping the resource attributes as span attributes anyways.

https://github.com/GoogleCloudPlatform/opentelemetry-operations-go/blob/182fb273d196283030190c6f8d62cca466edd48a/exporter/trace/trace_proto.go#L100-L128

Reply:
Doh